cdap-system-metrics/src/main/java/io/cdap/cdap/metrics/publisher/MetricsWritersMetricsPublisher.java

```
@@ -0,0 +1,118 @@
+ */
+public class MetricsWritersMetricsPublisher extends AbstractMetricsPublisher {
+  private static final Logger LOG = LoggerFactory.getLogger(MetricsWritersMetricsPublisher.class);
+  private final Map<String, MetricsWriter> metricsWriters;
+
+  @Inject
+  public MetricsWritersMetricsPublisher(MetricsWriterProvider writerProvider, CConfiguration cConf) {
+    this.metricsWriters = writerProvider.loadMetricsWriters();
+    initializeMetricWriters(this.metricsWriters, cConf);
+  }
+
+  private static void initializeMetricWriters(Map<String, MetricsWriter> metricsWriters, CConfiguration cConf) {
+    for (Map.Entry<String, MetricsWriter> entry : metricsWriters.entrySet()) {
+      MetricsWriter writer = entry.getValue();
+      // Metrics context used by MetricsStoreMetricsWriter only, which we don't use here
+      // So we can pass no-op context
+      DefaultMetricsWriterContext metricsWriterContext =
+        new DefaultMetricsWriterContext(new NoopMetricsContext(), cConf, writer.getID());
+      writer.initialize(metricsWriterContext);
+    }
+  }
+
+  /**
+   * Function that does the following:
+   * 1. Return {@link Exception}{@code e} if {@code collector} is null.
+   * 2. Return {@code collector} after adding a {@link Exception e} as a suppressed exception.
+   * to {@code collector} if {@code collector} isn't null.
+   *
+   * @param collector An {@link Exception} that is used to collect suppressed exceptions.
+   * @param e         An {@link Exception} to be added as a suppressed exception.
+   * @return
+   */
+  private static Exception addOrAssignException(Exception collector, Exception e) {
+    if (collector == null) {
+      return e;
+    }
+    collector.addSuppressed(e);
+    return collector;
+  }
+
+  @VisibleForTesting
+  private static void closeMetricWriters(Map<String, MetricsWriter> writers) throws IOException {
+    IOException exceptionCollector = null;
+    for (Map.Entry<String, MetricsWriter> entry : writers.entrySet()) {
```

It should be fine to log and ignore the errors in close.

---

Changes made to only log errors and not throw any exception.